libdolbyb/libdolbyb.3
'\" t
'\" The line above instructs most `man' programs to invoke tbl
'\"
'\" Separate paragraphs; not the same as PP which resets indent level.
.de SP
.if t .sp .5
.if n .sp
..
'\"
'\" Enclosure macro for examples
.de XE
.SP
.nf
.ft CW
..
.de XX
.ft R
.SP
.fi
..
.TH libdolbyb 3 "April 16, 2025" "libdolbyb"
.SH NAME
libdolbyb, a Dolby B encoding and decoding library
.SH SYNOPSIS
.B #include <dolbyb.h>
.na
.nh
.P
.B dolbyb_t dolbyb;
.P
.B void dolbyb_init(&dolbyb);
.P
.B char *dolbyb_start(&dolbyb);
.P
.B char *dolbyb_restart(&dolbyb);
.P
.B char *dolbyb_encode(&dolbyb, void *in, void *out, size_t nframes);
.P
.B char *dolbyb_decode(&dolbyb, void *in, void *out, size_t nframes);
.P
.B void dolbyb_free(&dolbyb);
.SH DESCRIPTION
.B libdolbyb
is a library to encode and decode audio applying or removing
Dolby\ B noise reduction for magnetic tape recording.
.TP
.B dolbyb_init
Fills in your \fIdolbyb_t\fP structure with default values.
.TP
.B dolbyb_start
Fills in parts of the \fIdolbyb\fP structure with values that depend on the
sample rate, number of channels and bit depth and calibrates the circuitry.
Between \fBdolbyb_init()\fR and \fBdolbyb_start()\fR you must set:
.RS
.TP
.B dolbyb.SmpSec
The sampling rate of the incoming (and outgoing) audio.
.PP
you can set:
.TP
.B dolbyb.NumChn
The number of channels in the audio: 1 or 2 with a default of 1.
.TP
.B dolbyb.BDepth
The size of each sample in bits: 8, 16 or 24.
8-bit unsigned and 24-bit signed don't yet work perfectly 
so 16 (signed) is the default.
.PP
and you may also set:
.TP
.B dolbyb.UpSamp
Digital filtering only works well if the sample rate is well above
the cutoff frequency of the filter. For Dolby\ B's sliding filter,
that frequency can be as high as 34kHz and this does not work well
if the sample rate is only 44.1Khz. To get around this, it upsamples
the audio to a higher rate when it passes through this filter.
By default, the upsample rate is set so that the upper sample rate
is at least 200Khz; upsampling can be switched off by using \fB\-u1\fR.
.TP
.B dolbyb.AllHig
If \fB\-h\fR is given, upsampling is used throughout the effect
from when the audio enters the effect to when is leaves.
As \fBdolbyb\fR's up/downsampling algorithm is simple 
(repeating and averaging samples) you may obtain higher quality results
by upsampling before \fBdolbyb\fR, using \fB\-u1\fR and
downsampling afterwards.
.TP
.B dolbyb.ThGndB
When a tape deck is encoding or decoding a magnetic tape,
it knows the signal level at the tape heads but with audio files
the maximum signal level may not accurately represent the tape's
maximum flux density (200nWb/m for cassette tapes), resulting in 
erroneous results. To overcome this, \fBThGndB\fR adjusts the volume level
at which the sliding filter reacts.
Its default value is 0.0, which assumes that the maximum amplitude
of the signal represents the maximum recording level on tape.
To begin with, when you have little idea of what level to use,
try a wide range of levels like 5, 10, 15 and 20.
If the result sounds muffled, the threshold is too low and
if it seems to have too much treble, the threshold is too high.
Once you know the approximate level, you can try more closely spaced levels
and listen carefully to find the best level possible.
Logic would suggest listening to where tracks fade out, to see if
the treble increases but this method doesn't seem to work well
and the best way seems to be to see how low the level can be set
before the results sound dull and muffled, then choose a level
a bit higher than this; you can just about hear the difference
between results that differ in threshold setting by about 2.
.TP
.B dolbyb.DecAdb
In decode mode, the program has to use trial and error to get the
right output sample values. This parameter sets how accurate it
needs to be before it is considered OK. A figure of 0.0\ Db would
mean an accuracy of about 1 sample value. The default is -5.0\ dB,
which is accurate to less than one sample value. A higher value
would be less accurate but it may make the decode faster; a lower
value would be more accurate but it may make the program slower.
.TP
.B dolbyb.FltTyp
The program originally simulated an analog circuit for a Dolby\ B
noise reducer. However, too much filtering in the side path was
altering the phase of the side path audio, which caused problems
when the side path was recombined with the main signal. Basically
signals don't add together very well if there is too much difference
in the phase.
.SP
There are now 4 filter modes with hopefully less of a phase change:
.RS
.TP
.B 1
is the original method.
.TP
.B 2
is a newer method that seems to work better than 1.
.TP
.B 3
is another rearrangement which in practice doesn't seem to
be any better than 1.
.TP
.B 4
seems to work best, hence it is the default mode.
.RE
.SP
\fBdolbyb_start()\fR can only fail (and return a non-zero value)
if you set parameters to out-of-range values or
if there is no memory available for the few buffers it needs to allocate.
.TP
.B dolbyb_restart
Can be called in the middle of an encoding or decoding session if you change
\f(CWdolbyb.{ThGndB,DecAdB,UpSamp,AllHig,FltTyp}\fR, to update internal values
that depend on them.
.TP
.B dolbyb_encode
Dolby B encodes a buffer of audio samples with 1, 2 or 3 bytes per sample,
with the samples for each channel interleaved, and writes the same number
of samples to the \fIout\fP sample buffer in the same format.
It cannot fail, believe it or not.
.TP
.B dolbyb_encode
Decodes a buffer of audio samples, the same way as \fBdolbyb_encode()\fP.
It also cannot fail.
.TP
.B dolbyb_free
Frees memory that was allocated by \fBdolbyb_start()\fP.
Calling it even if \fBdolbyb_start()\fP failed is safe.
.PP
The library is thread-safe and you can run multiple instances of it
simultaneously by having several \fBdolbyb_t\fP structures.
If they're all using the same number of channels, sample rate and bit depth,
you can simply copy the structure filled in by \fBdolbyb_start()\fP
and hand the different copies to the encoder and decoder functions,
in which case just make sure you only \fBdolbyb_free()\fP one of them.
.SH RETURN VALUE
All functions except dolbyb_init() and dolbyb_free() return NULL if they succeed
or a pointer to an error message if they fail.
.SH EXAMPLES
The following code uses \fBlibsndfile\fR to read and write
audio files. Put it in \fBdolbyb.c\fR and compile it with
.XE
cc -o dolbyb dolbyb.c -ldolbyb -lsndfile -lm
.XX
.XE
#include <stdlib.h>
#include <stdio.h>
#include <dolbyb.h>
#include <sndfile.h>

static void
usage(void)
{
  fprintf(stderr, "Usage: dolbyb [options] in.wav out.wav\n");
  fprintf(stderr, "-d   Decode a Dolby B-encoded recording (this is the default action)\n");
  fprintf(stderr, "-e   Encode instead of decoding\n");
  fprintf(stderr, "-uN  Upsample the low-pass filter N times\n");
  fprintf(stderr, "     (defaults to the right value for >=200kHz)\n");
  fprintf(stderr, "-h   Process absolutely everything at the upsampled rate\n");
  fprintf(stderr, "-tN  Set the threshold gain to N dB (default 0.0)\n");
  fprintf(stderr, "-aN  Set the decode accuracy to N dB (default -5.0)\n");
  fprintf(stderr, "-fN  Use filter type N (1-4, default 4)\n");
}

#define HANDFUL 1024   /* How many sample frames to read/write at a time */

int
main(int argc, char *argv[])
{
  dolbyb_t dolbyb;
  SNDFILE *sf_in, *sf_out;
  SF_INFO sfinfo;
  short *buf_in, *buf_out;
  sf_count_t n_read;
  int Encode = 0;

  dolbyb_init(&dolbyb);

  /* Process commandline arguments */
  argv++; argc--; /* Skip program name */
  while (argc > 0 && argv[0][0] == '-') {
    switch (argv[0][1]) {
    case 'd': Encode        = 0; break;
    case 'e': Encode        = 1; break;
    case 'h': dolbyb.AllHig = 1; break;
    case 'f': dolbyb.FltTyp = argv[0][1] - '0'; break;
    case 'u': dolbyb.UpSamp = atol(&argv[0][2]); break;
    case 't': dolbyb.ThGndB = atof(&argv[0][2]); break;
    case 'a': dolbyb.DecAdB = atof(&argv[0][1]); break;
    default: usage(); exit(1);
    }
    argv++; argc--;
  }
  if (argc != 2) { usage(); exit(1); }

  sf_in = sf_open(argv[0], SFM_READ, &sfinfo);
  if (sf_in == NULL) exit(1);
  sf_out = sf_open(argv[1], SFM_WRITE, &sfinfo);
  if (sf_out == NULL) exit(1);

  buf_in = calloc(sfinfo.channels * sizeof(short), HANDFUL);
  buf_out = calloc(sfinfo.channels * sizeof(short), HANDFUL);
  if (buf_in == NULL || buf_out == NULL) exit(1);

  dolbyb_init(&dolbyb);
  dolbyb.SmpSec = sfinfo.samplerate;
  dolbyb.NumChn = sfinfo.channels;
  dolbyb.BDepth = 16;
  if (dolbyb_start(&dolbyb)) exit(1);

  while ((n_read = sf_readf_short(sf_in, buf_in, (sf_count_t)HANDFUL)) > 0) {
    if (Encode) dolbyb_encode(&dolbyb, buf_in, buf_out, (size_t)n_read);
    else        dolbyb_decode(&dolbyb, buf_in, buf_out, (size_t)n_read);
    if (sf_writef_short(sf_out, buf_out, n_read) != n_read) {
      fprintf(stderr, "Write error on the output file.\n");
      break;
    }
  }
  sf_close(sf_out);
  sf_close(sf_in);
  dolbyb_free(&dolbyb);

  return 0;
}
.XX
.SH COPYRIGHT
Copyright 1991\-2015 Lance Norskog, Chris Bagwell and sundry contributors.
.SP
This library is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2
as published by the Free Software Foundation.
.SP
This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.
.SH AUTHORS
The many authors and contributors are listed in the README file that
is distributed with the source code.
.SH SEE ALSO
dolbyb(1), https://codeberg.org/martinwguy/libdolbyb


ChangeLog
Change History
--------------

This file contains a list of all changes starting after the release of
sox-11gamma.


sox_ng-14.6.0.2	2025-07-03
---------------

Bug fixes:

  o Fix use of lsx_valloc() in win32-unicode.c on MSYS2/MingW
  o Don't include test/samples in the distro tarball


sox_ng-14.6.0.1	2025-06-27
---------------

Bug fixes:

  o Fix compilation on MSYS2/MinGW with GCC 15.1
  o Fix compilation on MacOS X with older SDKs
  o Fix opusfile detection when using pkg-config
  o Fix compilation of libdolbyb
  o Fix compilation on Sortix
  o Fix compilation of libdolbyb on AIX 7.3


sox_ng-14.6.0	2025-06-17
-------------

New features:

  o Read NSP format files
  o Read MSP2K (Akai sampler) format files
  o Read and write AIFC a-law and u-law encodings
  o Copy AIFF's MARK and INST chunks more often
  o Copy AIFC's MARK and INST chunks
  o Read and write DSF format and read DFF and WSD formats
  o Add the DSD-related "dop" and "sdm" effects
  o Add reading of Ogg FLAC files (with ffmpeg)
  o SDS format is now autodetected
  o MP3 with CRC protection and MP2 formats are now autodetected
  o AIFF format now supports ID3 tags
  o The ID3 "TCOM" (Composer) tag is now supported
  o Add "softvol" effect, a software gain that avoids clipping and maybe rises
  o Make the 'v' and 'V' keys prefer to adjust softvol if it is active
  o Add "dolbyb" effect to decode and encode Dolby B noise reduction
  o synth: Add "vdelay" combine method to phase-modulate by the synth wave
  o Remove the limit on the number of echo/echos stages (was 7)
  o Remove the limit on the number of channels that phaser can handle (was 4)
  o Raise the limit on the number of chorus stages from 7 to 256
  o Remove unnecessary limits on arguments to echo,echos,chorus,phaser,flanger
  o wget, wget2 and curl are sought at runtime instead of at build time and
    ./configure --with-curl now just makes it prefer curl to wget.
  o Make "echos" run 20% faster
  o Make --multi-threaded the default
  o Error messages have been harmonized. The best was
    sox_ng WARN chorus: chorus: warning >>> gain-out can cause saturation
  o When curl or wget fail to fetch a URL, decode the exit status to a message
  o Don't spew the usage pages when there are invalid options
  o Add flow diagrams to --help-effect chorus/compand/earwax/echo/echos/phaser
  o Make --with-ffmpeg the default configuration

Bug fixes:

  o The "stats" effect's Bit-depth values have been revised to reflect
    what Bit-depth means and handle corner cases like -32768
  o Revise all effects' usage messages to the same format and more helpful
  o "echos" and "chorus" have never worked but now do what they say
  o "bend"'s logarithmic frequency curve is more closely logarithmic
  o Fix "silence" not to add a random selection of fragments at the end
  o Make "firfit" understand frequencies like "10k"
  o Read mono 8-bit MAUD files with an odd number of samples
  o Handle unknown odd-sized chunks in AIFF files
  o Detect and report read errors from short audio files
  o Detect and report write errors to audio files
  o When data read from a pipe is short, report stats for what there was
  o When an fatal error is detected, stop and exit non-zero in all cases
  o No formats or effects are experimental or deprecated any more
  o Multi-channel LADSPA plugins now work
  o The LD_LIBRARY_PATH change has been reverted (it bombed on Solaris)
  o Revert the "Unicode when writing id3 tags" patch which never worked
  o ffmpeg's ADX format is no longer autodetected (too many false positives)
  o Reject parameter values of NaN instead of dumping core
  o Correct the manual and make its mcompand example work

sox_ng-14.5.0.3	2025-01-05
---------------

Bug fixes:

  o Make reading from ffmpeg work on Unix too.
    popen(rb) returns NULL with glibc (!)

sox_ng-14.5.0.2	2024-12-23
---------------

Bug fixes:

  o Make reading from ffmpeg work by popen()ing "rb"

sox_ng-14.5.0.1	2024-12-23
---------------

New features:

  o Add Unicode support on Windows for filenames and command-line arguments

Bug fixes:

  o Make it build for Windows using MXE

sox_ng-14.5.0	2024-12-06
-------------

New features:

  o Add effect speexdsp with automatic gain control and noise reduction
  o Auto-detect MP3 files
  o If ./configured --with-ffmpeg and ffmpeg is installed, autoread formats:
    3g2, 3gp, aac, ac3, adts, adx, ape, apm, aptx, argo_asf, asf, ast, avi,
    dfpwm, dts, dvd, eac3, f4v, flv, gxf, ism, kvag, m4a, m4v, mkv, mlp,
    mp4, mpeg, mpegts, mxf, nut, oga (Ogg with FLAC data), ra, rm, rso, sbc,
    smjpeg, spdif, speex, svcd, tta, vag, vcd, vob, webm, wma, wsaud, wtv
    and yet more with -t ffmpeg
  o Add "stat -a" to give the average power spectrum
  o Add "spectrogram -n" flag to normalize its brightness
  o Add "spectrogram -L" flag to give a logarithmic frequency axis
  o Add "spectrogram -R" flag to specify the frequency range
  o Raise spectrogram's height limit from 8193 to 200000 pixels
  o Lower the minimum speed of the flanger effect from 0.1 to 0.01 Hz
  o Make combine effects work when there's a single file
  o riaa: Add 192kHz sample rate
  o sphere format: Support ALAW encoding
  o SD2 format: Support resource forks
  o ID3 tags: Support unicode when writing
  o WAV files: Read when the number of valid bits is less that the sample size
  o Use posix_fadvise to increase readahead and double its speed
  o Use FFTW to make non-2^n-size spectrograms a hundred times faster
  o Resize Linux pipe buffers to make multi-threaded effects 10-80% faster
  o Reduce sox -t pulseaudio --[input-]buffer latency from up to 2 secs to low
  o Enable building to fetch URLs with curl instead of wget
  o Be able to read files that are still being written by another process
  o Make "make check" run the regression test suite
  o Drop sox_version_info_t's "time" element to get reproducible builds
  o Remove the undocumented and useless "divide" effect

libdolbyb/libdolbyb.c
/*
 * libdolbyb.c - API to libdolbyb
 *
 * Copyright (C) 2025 Martin Guy <martinwguy@gmail.com>
 * based on dolbybcsoftwaredecode by Richard Evans 2018.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation. See COPYING for details.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "dolbyb.h"
#include "Param.h"
#include "Mixers.h"
#include "Calibrate.h"
#include "SidePath.h"
#include "FindOutSmp.h"

#include "../src/soxconfig.h"  /* for WORDS_BIGENDIAN */

#include <stdlib.h>  /* for free() */
#include <string.h>  /* for memset() */

/* Set default values */
void dolbyb_init(dolbyb_t *Param)
{
  memset((void *)Param, 0, sizeof(*Param));

  /* Set default values */
  Param->NumChn = 1;
  Param->BDepth = 16;
  Param->FltTyp = 4;
  Param->DecAdB = -5.0;
  Param->ThGndB = 0.0;

  /* Set initial values */
  Param->FETGVt = 75000*(int64_t)100000;
}

/* Update things that depend on changed values of the
 * settable parameters ThGndB, DecAdB, UpSamp, AllHig, FltTyp.
 * It will also readjust for changed BDepth and SmpSec but
 * I'm not sure what that means in the context of a session.
 */
char *dolbyb_restart(dolbyb_t *Param)
{
  return dolbyb_start(Param);
}

static char *SecondInit(dolbyb_t *Param)
{
  char *err;

  /* More initialising after the input wave header has been read,  */
  /* so the sample rate and number of samples is now known.        */

  Param->DecAMX = ParamConvertDb(Param->DecAdB);
  Param->ThGain = ParamConvertDb(Param->ThGndB);

  /* Set up sample rates */
  if (Param->UpSamp < 0) return "Upsampling cannot be negative";
  if (Param->UpSamp > 0)
    Param->UpSmp = Param->UpSamp;
  else {
    Param->UpSmp = 1;
    while (Param->UpSmp * Param->SmpSec < 200000L)
      Param->UpSmp++;
  }
  if (Param->AllHig)
    Param->InUS = Param->UpSmp;
  else
    Param->InUS = 1;

  /* Initialise side path routines */
  if ((err = SidePathInit(Param))) return err;
  FindOutSmpInit(Param);
  return NULL;
}

/* Do initialisation that depends on parameter changes */
char *dolbyb_start(dolbyb_t *Param)
{
  char *err;

  /* Check validity of parameters */

  if (Param->SmpSec <= 0)
    return "Did you forget to set dolbyb.SmpSec before calling dolbyb_start()?";

  if (Param->NumChn < 1 || Param->NumChn > 2)
    return "libdolbyb can only process mono and stereo audio";

  /* Check filter type is valid */
  if (Param->FltTyp < 1 || Param->FltTyp > 4)
    return "dolbyb.FltTyp must be from 1 to 4";

  switch (Param->FltTyp) {
  case 1: Param->FETSVt = ParamFETSVt1; Param->SidAmp = ParamSidAmp1; break;
  case 2: Param->FETSVt = ParamFETSVt2; Param->SidAmp = ParamSidAmp2; break;
  case 3: Param->FETSVt = ParamFETSVt3; Param->SidAmp = ParamSidAmp3; break;
  case 4: Param->FETSVt = ParamFETSVt4; Param->SidAmp = ParamSidAmp4; break;
  }

  /* Set multiplier for wave input and divider for wave output */
  switch (Param->BDepth) {
  case 8:  Param->SmpMux = ParamSMux08; break;
  case 16: Param->SmpMux = ParamSMux16; break;
  case 24: Param->SmpMux = ParamSMux24; break;
  default:
    return "dolbyb.BDepth must be 8, 16 or 24";
  }

  if ((err = SecondInit(Param))) return err;

  Calibrate(Param);

  return NULL;
}

char *dolbyb_encode(dolbyb_t *Param, void *in, void *out, size_t nframes)
{
  unsigned char *inp = in;
  unsigned char *outp = out;
  uint16_t Chn, UpCnt;
  int64_t SidSmp, TotSmp, OutSmp;
  size_t SmpCnt;
  int64_t MaxSamp, SubSamp;
  uint16_t NumByt;
  int32_t MaxVal, MinVal, AddVal;

  switch (Param->BDepth) {
  case 8:  NumByt = 1; MaxSamp = -1; SubSamp = 128; 
           MaxVal = 127; MinVal = -128; AddVal = 128;
           break;
  case 16: NumByt = 2; MaxSamp = 32767; SubSamp = 65536L;
           MaxVal = 32767; MinVal = -32768L; AddVal = 65536L;
           break;
  case 24: NumByt = 3; MaxSamp = 8388607L; SubSamp = 16777216L;
           MaxVal = 8388607L; MinVal = -8388608L; AddVal = 16777216L;
           break;
  default:
	   /* Shut the compiler warnings up */
           NumByt = MaxSamp = SubSamp = 0;
           MaxVal = MinVal = AddVal = 0;
           return "BDepth is not 8/16/24. Did you set it before calling dolbyb_start()?";
  }

  for (SmpCnt = 0; SmpCnt < nframes; SmpCnt++) {
    for (Chn = 1; Chn <= Param->NumChn; Chn++) {
      int64_t SmpVal = 0; /* Shut compiler warning up */
      union {
        unsigned char c[4];
	signed int i;
      } u;

      /* Get input */
      switch (NumByt) {
      case 1: SmpVal = (int64_t)(inp[0]);
              break;
      case 2: u.i = 0;  u.c[0] = inp[0]; u.c[1] = inp[1];
              SmpVal = u.i;
              break;
      case 3: u.i = 0;  u.c[0] = inp[0]; u.c[1] = inp[1]; u.c[2] = inp[2];
              SmpVal = u.i;
              break;
      }
      inp += NumByt;
      if (SmpVal > MaxSamp) SmpVal -= SubSamp;
      SmpVal *= Param->SmpMux;

      /* If upsampling at the start then the same sample */
      /* is processed several times */
      TotSmp = 0;
      for (UpCnt = 1; UpCnt <= Param->InUS; UpCnt++) {
	/* Add audio from side path */
	SidSmp = SidePath(Param, SmpVal, Chn);
	TotSmp += MixersEncode(SmpVal, SidSmp);
      }
      OutSmp = (TotSmp / Param->InUS) / Param->SmpMux;

      /* Check for sample value out of range */
      if (OutSmp > MaxVal)
	SmpVal = MaxVal;
      else if (OutSmp < MinVal)
	SmpVal = MinVal;
      else
	SmpVal = OutSmp;
      /* Convert negative values */
      if (SmpVal < 0 || Param->BDepth == 8)
	SmpVal += AddVal;

      /* Send as output bytes */

      switch (NumByt) {
      case 1: *outp++ = SmpVal & 0xff;
              break;
#ifndef WORDS_BIGENDIAN
      case 2: *outp++ = SmpVal & 0xff;
              *outp++ = (SmpVal >> 8) & 0xff;
              break;
      case 3: *outp++ = SmpVal & 0xff;
              *outp++ = (SmpVal >> 8) & 0xff;
              *outp++ = (SmpVal >> 16) & 0xff;
#else
      case 2: *outp++ = (SmpVal >> 8) & 0xff;
              *outp++ = SmpVal & 0xff;
              break;
      case 3: *outp++ = (SmpVal >> 16) & 0xff;
              *outp++ = (SmpVal >> 8) & 0xff;
              *outp++ = SmpVal & 0xff;
#endif
              break;
      }
    }
  }
  return NULL;
}

char *dolbyb_decode(dolbyb_t *Param, void *in, void *out, size_t nframes)
{
  unsigned char *inp = in;
  unsigned char *outp = out;
  uint16_t NumChn, UpCnt, NumByt;
  int64_t TotSmp;
  int64_t MaxSamp, SubSamp, MinVal, MaxVal, AddVal;
  size_t SmpCnt;

  switch (Param->BDepth) {
  case 8:  NumByt = 1; MaxSamp = -1; SubSamp = 128; 
           MaxVal = 127; MinVal = -128; AddVal = 128;
           break;
  case 16: NumByt = 2; MaxSamp = 32767; SubSamp = 65536L;
           MaxVal = 32767; MinVal = -32768L; AddVal = 65536L;
           break;
  case 24: NumByt = 3; MaxSamp = 8388607L; SubSamp = 16777216L;
           MaxVal = 8388607L; MinVal = -8388608L; AddVal = 16777216L;
           break;
  default:
	   /* Shut the compiler warnings up */
           NumByt = MaxSamp = SubSamp = 0;
           MaxVal = MinVal = AddVal = 0;
           return "BDepth is not 8/16/24. Did you set it before calling dolbyb_start()?";
  }

  /* Process samples */
  for (SmpCnt = 0; SmpCnt < nframes; SmpCnt++) {
    for (NumChn = 1; NumChn <= Param->NumChn; NumChn++) {
      int64_t SmpVal, OutSmp;

      /* Get input */
      switch (NumByt) {
      case 1: SmpVal = (int64_t)(inp[0]);
              break;
#ifndef WORDS_BIGENDIAN
      case 2: SmpVal = (int64_t)((int32_t)inp[0] | ((int32_t)inp[1] << 8));
              break;
      case 3: SmpVal = (int64_t)((int32_t)inp[0] | ((int32_t)inp[1] << 8) | ((int32_t)inp[2] << 16));
#else
      case 2: SmpVal = (int64_t)((int32_t)inp[1] | ((int32_t)inp[0] << 8));
              break;
      case 3: SmpVal = (int64_t)((int32_t)inp[2] | ((int32_t)inp[1] << 8) | ((int32_t)inp[0] << 16));
#endif
              break;
      }
      inp += NumByt;
      if (SmpVal > MaxSamp) SmpVal -= SubSamp;
      SmpVal *= Param->SmpMux;

      /* If upsampling at the start then the same sample */
      /* is processed several times */
      TotSmp = 0;
      for (UpCnt = 1; UpCnt <= Param->InUS; UpCnt++) {
	/* Set and process output */
	(void) FindOutSmp(Param, SmpVal, NumChn);
	TotSmp += MixersDecode(SmpVal, SidePathUpdate(Param, NumChn));
      }
      OutSmp = (TotSmp / Param->InUS) / Param->SmpMux;

      /* Check for sample value out of range */
      if (OutSmp > MaxVal)
	SmpVal = MaxVal;
      else if (OutSmp < MinVal)
	SmpVal = MinVal;
      else
	SmpVal = OutSmp;
      /* Convert negative values */
      if (SmpVal < 0 || Param->BDepth == 8)
	SmpVal += AddVal;

      /* Send as output bytes */
      switch (NumByt) {
      case 1: *outp++ = SmpVal & 0xff;
              break;
#ifndef WORDS_BIGENDIAN
      case 2: *outp++ = SmpVal & 0xff;
              *outp++ = (SmpVal >> 8) & 0xff;
              break;
      case 3: *outp++ = SmpVal & 0xff;
              *outp++ = (SmpVal >> 8) & 0xff;
              *outp++ = (SmpVal >> 16) & 0xff;
#else
      case 2: *outp++ = (SmpVal >> 8) & 0xff;
              *outp++ = SmpVal & 0xff;
              break;
      case 3: *outp++ = (SmpVal >> 16) & 0xff;
              *outp++ = (SmpVal >> 8) & 0xff;
              *outp++ = SmpVal & 0xff;
#endif
              break;
      }
    }
  }
  return NULL;
}

void dolbyb_free(dolbyb_t *Param)
{
  if (Param->HPF2SetValsPotTab) {
    free(Param->HPF2SetValsPotTab); Param->HPF2SetValsPotTab = NULL;
  }
  if (Param->HPF2SetValsAlpTab) {
    free(Param->HPF2SetValsAlpTab); Param->HPF2SetValsAlpTab = NULL;
  }
  if (Param->HPF2SetValsFLTATb) {
    free(Param->HPF2SetValsFLTATb); Param->HPF2SetValsFLTATb = NULL;
  }
  if (Param->SetGateAttTab) {
    free(Param->SetGateAttTab); Param->SetGateAttTab = NULL;
  }
  if (Param->SetGateAlpTab1) {
    free(Param->SetGateAlpTab1); Param->SetGateAlpTab1 = NULL;
  }
  if (Param->SetGateAlpTab2) {
    free(Param->SetGateAlpTab2); Param->SetGateAlpTab2 = NULL;
  }
}
